README.md
ace
===

Lightweight FE/BE Framework. Provides tools for launching prototype websites.

[Visit Demo Page](http://ace.fabfitfun.com/demo)


### Install via git
```bash
git clone git@github.com:fluffybunnies/ace.git path-to-project
```
### Install via npm
```bash
mysite=./mysite; if [ ! -d "$mysite" ]; then mkdir -p "$mysite"; fi; npm install --prefix /tmp aced && cp -ri /tmp/node_modules/aced/* "$mysite" && cd "$mysite" && ls -l
```
### Or, deploy directly to an ubuntu instance with [Sire](https://github.com/fluffybunnies/sire)
```bash
# Deploy sire to remote server:
./sire/index.sh _deploy
# Deploy ace to remote server:
./sire/signal.sh ace
```


### Front End - UI

##### Instagram Gallery
- Plug and play with a simple gallery id
- Or configure an api mapping to generate dynamic galleries
```html
<script type="text/ace-instagram">{
	query: 'users/227962011/media/recent'
}</script>
```
```javascript
ace.ui.widgetize('instagram',$('.gallery01'),{
	query: 'users/%userId%/media/recent'
	,fetch: {
		userId: ['users/search?q=markthegonzales',function(res){
			return res.data[0].id;
		}]
	}
	,type: 'squares med'
	,num: 10
	,hoverFadeIn: true
	,shadbox: true
});
```

##### Twitter Feed
- Horizontal or vertical ticker animation
```html
<script type="text/ace-twitter">{
	screenName: 'jewelmint'
	,scroll: 'x'
}</script>
```
```javascript
ace.ui.widgetize('twitter',$('.sidebar-social'),{
	screenName: 'pandaexpress'
	,numGet: 10
	,numShow: 3
	,scroll: 'y'
	,scrollDelay: 8000
	,scrollSpeed: 1000
	,type: 'sidebar'
});
```

##### Simple Carousel
```html
<script type="text/ace-carousel">{
	imgs: [
		'/imgs/fashion_jewelry.jpg'
		,'/imgs/three-stones-engagement-ring.jpg'
		,'/imgs/Chanel_Wallet_Chriselle1.jpg'
		,'/imgs/Peplum_Leather_zara_top_Valentino_rockstud.jpg'
		,'/imgs/summer-arm-candy.jpg'
		,'/imgs/boho-chic-2.jpg'
	]
}</script>
```
```javascript
ace.ui.widgetize('carousel',$('.nyfw-2015'),{
	imgs: albumNYFW2015
	,dims: '195x148'
	,speed: 300
	,shadbox: true
});
```

##### Chat
See [demo page](http://ace.fabfitfun.com/demo) and web/assets/ace.chat.js

Uses [arrrr](https://www.npmjs.com/package/arrrr) module for backend

##### Tooltips
- Place smart tooltips on any html element
- Intelligent positioning adapts to custom css styles and always remains within viewport
- Activated by hover, click, or any other custom event
- Supports tip activation groupings
```javascript
ace.tooltip($('#look-at-me'),{
	content: 'I do cool things!'
});

// or...

ace.tooltip($('#me-too'),{
	content: 'Click here for lazers'
	,offset: 10 // null = width of arrow
	,pos: 'top' // top bot left right
	,fixed: false
	,group: null // tips with the same group will show/hide at the same time
	,classes: 'tooltip-mkII'
	,showEvt: 'mouseover'
	,hideEvt: 'mouseout'
});
```

##### ShadBox
- Bring images to life with this shadow box modal
- Makes load-time look good, with options to customize the animation
- Plugs directly into other ace modules such as the photo carousel and instagram gallery via a simple on/off option
```javascript
ace.shadbox('/albums/2015/some.sweet.image.jpg');

// or...

ace.shadbox('/albums/2014/hard.rock.25may2014.jpg',{
	viewport: {
		padding: {x:.05, y:.05}
	}
	,anim: {
		fadeSpeed: 100
		,delay: 300
		,contentExpandSpeed: 300
		,contentFadeInSpeed: 300
	}
});
```

##### Pop
- Use as a simple alert dialog
- Or extend with versatile classes and bindings
```
ace.pop('An email has been sent with a link to reset your password.');

// or...

ace.pop({
	header: 'Are you sure?'
	,body: 'Clicking Ok will delete everything you own.'
	,btns: [
		['cancel','Cancel']
		,['ok','Ok']
	]
	,exit_btn: true
	,position: 'fixed'
	,true_center: false
}).on('ok',function(){
	forfeitTheWorld();
});
```

##### Loader
- Loader overlay animation remains visible while "up()"s > "down()"s
```
ace.loader.up()
doSomethingAsync(function(){
	ace.loader.down()
})
ace.loader.up()
doSomethingElseAsync(function(){
	ace.loader.down()
})
```

##### Highlight
- Animate background color to bring attention to $element(s)
- Defaults to yellow fadeout over 1 second
- Use case: Highlight cell in datatable that was just updated
```
var $rowIJustUpdated = $('table tr.num-15');
ace.highlight($rowIJustUpdated).find('td.info').html(newValue);

$('body').addClass('flashMeGreen');
ace.highlight($('.flashMeGreen'), {
	start:'00ff00'
	,duration: 1500
	,framerate: 15
})
```

#### Resource
- On-demand versioned asset loading
- Think async requirejs lite with packs and css
```
// simple:
ace.resource.fetch('https://s3.example.com/assets/SNotifs.js');

// with callback:
ace.resource.fetch('https://s3.example.com/assets/SNotifs.js',function(){
	console.log('js and css loaded');
})

// multiple
ace.resource.fetch([
	'//s3.example.com/assets/Leaderboards.js'
	,'//s3.example.com/assets/Leaderboards.css'
],function(){
	console.log('everything loaded');
})

// split cbs
ace.resource.fetch([
	'//s3.example.com/assets/SActivity.pv2.js'
	,'//s3.example.com/assets/SActivity.pv2.css'
],function(){
	console.log('js loaded')
	letsStartFunctionalizing()
},function(){
	console.log('css loaded')
	renderThings()
});

// packs
ace.resource.fetch('pack:social')
```

#### Req
- Shorthand for api calls
```
ace.req('twitter/feed',function(err,data){
	if (err) return console.log(err)
	populateTwitterFeed()
})

ace.req('comments','post',{
	body: 'Thnks!'
})
```





### Front End - Utils
```javascript
// 'ace.util.'+Object.keys(ace.util).sort().join('\nace.util.')
ace.util.arrayFilter
ace.util.capitalize
ace.util.deleteCookie
ace.util.deobfu
ace.util.escapeHtml
ace.util.escapeRegEx
ace.util.formatDate
ace.util.formatInteger
ace.util.formatPlace
ace.util.formatTimeAgo
ace.util.getCookie
ace.util.getImageToWindowFit
ace.util.getParameterByName
ace.util.getViewportScrollY
ace.util.hash
ace.util.isEmptyObject
ace.util.isFullyWithinViewport
ace.util.isWithinViewport
ace.util.obfu
ace.util.onTouchDevice
ace.util.padZ
ace.util.parseCookies
ace.util.rand
ace.util.removeClassWithPrefix
ace.util.replaceAll
ace.util.setCookie
ace.util.setUniqueClassVal
ace.util.stdErrAlert
ace.util.strToClass
ace.util.trueDim
```


### Front End - Bus
All widgets (and the global ace.bus) extend AceBase - a standard event emitter with a "ready" implementation
```javascript
// ready() is ambivalent to whether or not the event has already been triggered
ace.bus.ready('something-happened',function(){
	console.log("Let's do something once!");
});
ace.bus.on('something-happened',function(){
	console.log("Let's do something everytime!");
});
ace.bus.trigger('something-happened');
ace.bus.trigger('something-happened');
/* ...
Let's do something once!
Let's do something everytime!
Let's do something everytime!
*/
```


### Back End - API
Basic API skeleton. Includes a few routes as examples:

##### Smile
http://ace.fabfitfun.com/ace/api/demo/smile

##### Send email with attachments
http://ace.fabfitfun.com/ace/api/demo/email-csv-php?email_to=alec@luckygroupinc.com



### Back End - Helpers

##### SES
Amazon SES wrapper. Super simple interface makes it easy to send emails with attachments.
```php
Ses::send(array(
	'to' => array('trinkledinkle@example.com'),
	'from' => 'stinklewinkle@gmail.com',
	'reply_to' => 'funky.munky@example.com',
	'bcc' => 'rufus@example.com',
	'subject' => 'Kali M lent you 2 loyalty points!',
	'message' => '<em>ssssssup</em>',
	'type' => 'html',
	'attachment' => WEBROOT.'/saved/secret_list.csv',
));
```

##### Protect
Contains basic methods to help secure your prototype website.
```php
Protect::preventBruteForce($numRequests=4, $perSeconds=2000, $usePath=true, $bank=50)
```



### Back End - Utils
```php
Ace::getConfig
	Request value loaded from config.php / config.local.php
Ace::g
	Example:
		if (Ace::g($_POST,'param') == 'abc') ...
		instead of
		if (isset($_POST['param']) && $_POST['param'] == 'abc') ...
	Example:
		$host = Ace::g($_SERVER,array('HOST_NAME','SERVER_NAME'));
		instead of
		$host = isset($_SERVER['HOST_NAME']) ? $_SERVER['HOST_NAME'] : isset($_SERVER['SERVER_NAME']) ? $_SERVER['SERVER_NAME'] : null;
Ace::onHttps
	Works behind ELB using HTTP_X_FORWARDED_PROTO
Ace::enforceHttps
	Redirect to https://self
Ace::vres
	Optimize asset caching by marking url with file-modified-time
Ace::e
	For debugging: exit with $message
Ace::isAssoc
	Returns true if array does not have continuously ascending numeric keys
Ace::varDump
	HTML-formatted alternative to var_dump
Ace::putDeep
	Insert into an array without checking isset()
	Example:
		$house = array(
			'bedroom' => array('bed','dresser'=>array('pants')),
		);
		Ace::putDeep($house,'kitchen>oven','bread');
		Ace::putDeep($house,'bedroom>dresser[]','shirt');
		Ace::varDump($house);
Ace::onMobile
	Uses list from http://detectmobilebrowsers.com/ against HTTP_USER_AGENT
Ace::clientIp
	Works behind ELB and other proxies
Ace::curlGet
	curlGet( $url [, $params [, $curlOpts ]] )
Ace::curlPost
	curlPost( $url [, $params [, $curlOpts [, $urlEncodedParams = false ]]] )
	$urlEncodedParams mimicks application/x-www-form-urlencoded as opposed to the default multipart/form-data
Ace::curlDelete
	curlDelete( $url [, $params [, $curlOpts ]] )
Ace::setAssetHeadersForFilename
	For use when loading a file via php instead of webserver
Ace::sphericalDistance
	Calculate the distance between 2 lat/lng points
Ace::strToTime
	Enforce app-consistent timezone reference
Ace::strToTimeUtc
	Enforce app-consistent timezone reference
Ace::date
	Enforce app-consistent timezone reference
Ace::dateUtc
	Enforce app-consistent timezone reference
Ace::dbTime
	Mysql date format. Same as Ace::date('Y-m-d H:i:s', $time);
Ace::dbTimeUtc
	Mysql date format. Same as Ace::dateUtc('Y-m-d H:i:s', $time);
Ace::time
	Allows overriding return value of `Ace::time()` for unit tests
Ace::aceTmz
	Enforce app-consistent timezone reference
```


### Back End - Node
If you deployed using Sire, node will be installed with the latest stable version.

##### Send email with attachments
```javascript
node ./bin/demo-emailCsv.js --emailTo='alec@luckygroupinc.com' --emailFrom='acquiremint@beachmint.com'
```

##### Utils
```javascript
// node -e "console.log('ut.'+Object.keys(require('./lib/ut.js')).sort().join('\nut.'))"
ut.dateDiff
ut.dbTime
ut.fileTime
ut.flipObjKeyVals
ut.getFirstChild
ut.getFirstKey
ut.isNumeric
ut.padZ
ut.pluckFromArray
ut.prettyTime
ut.rand
ut.replaceAll
ut.spawn
ut.stats
ut.superTrim
ut.trim
```


### Health Check
http://ace.fabfitfun.com/hc


### Status Page
http://ace.fabfitfun.com/id


### To Do
- Fix `ace.ui.instagram` (instagram api forced update)
- Import relevant bin/s (e.g. create_dto.php)
- Import magic adapter + dao methods
- Import Ace + bootshell + method comments + etc upgrades
	- Don't simply copy+paste work done in Ace::normalizePostData, unit test normal www-form-encoded post from node, maaaybe fall back to the json obj encapsed form
	- e.g. DaoSqlAbstract::makePivotWhere
- Import Eav
- Convert ace.highlight.js to ace.jq.highlight.js
	- Review current state (escaped globals, memory+proc benchmark, etc)
- Create example route to demonstrate DaoSqlAbstract
- Prepend semicolons to ace modules that begin with "(" to allow for alternate builders like gulp
- Add: SocialShare, flashUI, Lights, et al
- MVC-type routes separate from API
	- Import home.php
- Implement basic cache layer with memcache driver
	- Start with App::getAppVersion()
- Generate ace.min.js/ace.min.css on start + post-gitsync / file change
- Unit tests
- More descriptive error messages for missing initial configs






CHANGELOG.md
## 0.1.2 (2017-02-28)
[https://github.com/fluffybunnies/ace/compare/ddb040d062ec8ba616565d658442f67691e34efd...355359fee20660dc289e93ea3adac10e3f8fd651](diff)

Features

	- New javascript `ace.util` utility methods: `isEmptyObject`, `removeClassWithPrefix`, `replaceAll`, `setUniqueClassVal`, `stdErrAlert`
	- New javascript `ace` modules: `polyfill`, `stickyFormVals`
	- New node `lib/ut' utility methods: `dbTime`, `flipObjKeyVals`, `isNumeric`, `pluckFromArray`, `rand`, `replaceAll`
	- New php `Ace` utility methods: `time`, `dbTime`, `dbTimeUtc`
	- `Ace::g()` now handles objects as well as arrays
	- Set "content-type: application/json" header for non-jsonp POST api responses
		- Prevents some XSS attacks
		- Previously only set for GET requests
		- Warning: this may affect POSTs through iframes in some browsers
	- Implemented late static binding in Ace Exceptions for extensibility


Bugs

	- Fix "Corrupted log file" exception thrown on small log files generated by `Protect` Helper
	- Minor bug fixes to: `ace.tooltip`
	- Normalize value of php `REQUEST_METHOD` and `REQUEST_PATH` constants across common environments



## 0.1.1 (2015-07-25)
[https://github.com/fluffybunnies/ace/compare/3de3fbf0e85d2737c278ced48c42806134223ae3...ddb040d062ec8ba616565d658442f67691e34efd](diff)

Features

 - Update ace php autoloader to accomodate other custom autoloaders


Bugs

 - Minor edge-case fixes to: `ace.highlight`, `ace.tooltip`, `ace.loader`



## 0.1.0 (2015-07-25)
[https://github.com/fluffybunnies/ace/compare/519a838cce7121b766599bd529750cf1a08a5245...3de3fbf0e85d2737c278ced48c42806134223ae3](diff)

Features

	- Bumped to 0.1.0: New API scaffolding, paving the path for v1.0.0
		- Design pattern: CAO( Controller > Adapter > DAO < DTO )
	- New to `\ace\Ace`
		- `setAssetHeadersForFilename()`, `sphericalDistance()`, date/time helpers
	- Codified Exceptions	
	- Routes split by request method
	- New to `ace.js`
		- `ace.util.formatDate`
		- `ace.getAppVersion`
		- `ace.log`
	- `ace.loader.js`
		- No more finickyness re overlapping loader animations
	- `ace.highlight.js`
		- Try it, it's fun. Probably turn it into a jQ extension later
	- `ace.resource.js`
		- On-demand versioned asset loading. Think require with packs and css.
	- `ace.req.js`
		- Wrapper for calling api: `ace.req('comments','post',{msg:'sup!'})`
	- New shorthand: `ace.pop('Message body', 'Message header')`
	- Some new button stylez
	- Node webserver: `test_php_server.js`


Bugs

	- Patched XSS and CSRF vulnerabilities
	- Fixed issue with the event bus' `ready()` implementation
	- Fixed tooltip y positioning when !opts.fixed



## 0.0.21 (2015-06-30)

Features

	- +ace.utils: `isWithinViewport()`, `isFullyWithinViewport()`, `hash()`, `obfu()`, `deobfu()`, `deleteCookie()`, `padZ()`
	- +ace.fb.js
	- Improved performance for `ace.util.formatInteger()`
	- Reorganized libs


Bugs

	- Fixed bug with `ace.ui.twitter` when metadata was missing from tweets



## 0.0.20 (2015-04-09)

Features

	- Improved detail in README.md
	- Hyperlink media urls in Twitter feed


Bugs

	- Fixed encoding method for Twitter OAuth



## 0.0.16 (2015-02-22)

Features

	- + Ability to query Instagram by screename instead of numeric id


Bugs

	- Handle missing caption in Instagram api response



## 0.0.15 (2015-02-20)

Features

	- New helper Protect. First method preventBruteForce() provides protection against DOS or brute-force attacks against specific routes (e.g. wp-login.php for wordpress).


Bugs

	- Helper SES was broken when sending normal emails due to lack of backwards compatibility with new AWS SDK.



## 0.0.12 (2014-12-06)

Features

	- Option for more intelligent ace.util.capitalize when passed a proper name


web/demo.php
<?php
if (!defined('WEBROOT'))
	exit;
use \ace\Ace;
?><!doctype html>
<html>
<head>
<meta charset="utf-8" />
<title>Ace | Paradigm</title>
<link rel="stylesheet" type="text/css" href="/assets/reset.css" />
<!--<script src="/assets/jquery.1.8.min.js"></script>-->
<script src="/assets/jquery.1.11.1.min.js"></script>
<link rel="stylesheet" type="text/css" href="/ace.css" />
<script src="/ace.js"></script>
<script>

<?php if (Ace::getConfig('facebookAppId')): ?>
ace.fb.loadSdk({
	appId: '<?php echo Ace::getConfig('facebookAppId'); ?>'
	,version: 'v2.1'
	,xfbml: false
});
<?php endif; ?>

$(function(){
	$('.ace-smile').each(function(){
		var $el = $(this)
			,times = +$el.attr('x-times') || 1
			,num = 0
			,getSmile
		;
		getSmile = function(){
			ace.req('demo/smile',function(err,data){
				if (err) {
					console.log('ERROR','demo/smile',err,data);
					if (!num)
						return $el.html('api error: '+err.error);
				} else {
					var $face = $('<span class="ace-smile-face">'+data+'</span>').css({
						position: 'relative'
						,top: '-100px'
						,opacity: 0
					})
					num ? $el.children('span').eq(ace.util.rand(0,$el.children('span').length-1)).after($face) : $el.append($face);
console.log('INIT WIDTH',ace.util.trueDim($face).w); // #dont remove yet
					$face.css('margin','0 -'+ace.util.trueDim($face).w/2+'px');
					setTimeout(function(){
						$face.css({
							transition: 'margin 1s, top 1s, opacity 1s'
							,'-webkit-transition': 'margin 1s, top 1s, opacity 1s'
							,margin: '0 0.5em'
							,top: 0
							,opacity: 1
						});
						if (($el.children('span').length-1)%3 == 0) {
							setTimeout(function(){
								ace.highlight($face, {start:'ff69b4', duration:800});
							},1100);
						}
					},100);
				}
				if (++num < times)
					setTimeout(getSmile,2000);
			});
		};
		getSmile();
		$el.on('click','.ace-smile-face',function(){
			var $face = $(this);
console.log('SUP',ace.util.trueDim($face,true)); // #dont remove yet
			$face.css({
				opacity: 0
				,margin: '0 -'+ace.util.trueDim($face).w/2+'px'
			});
			getSmile();
			//ace.loader.up();
			setTimeout(function(){
				//ace.loader.down();
				$face.remove();
			},1000);
		});
	});

	var $acePopExample = $('#ace-pop-example').bind('click',function(e){
		e.preventDefault();
		ace.pop({
			header: 'Are you sure?'
			,body: 'Clicking Ok will close this window.<br /><br />...So will clicking Cancel.'
			,btns: [
				['cancel','Cancel']
				,['ok','Ok']
			]
		}).on('ok',function(){
			ace.fireworks.fire();
		});
	});
	ace.tooltip($acePopExample, {
		content: 'I has speech!'
		,pos: 'right'
		,offset: 10
	});

	var $aceTooltipExample = $('#ace-tooltip-example').bind('click',function(){return false;});
	ace.tooltip($aceTooltipExample, {
		content: 'Look at me!'
		,pos: 'top'
		,offset: 10
		,group: 1
	});
	ace.tooltip($aceTooltipExample, {
		content: 'I can have multiple tooltips attached.<br />And they always stay on-screen!'
		,pos: 'bot'
		,offset: 10
		,group: 1
	});

	var $fbComments = $('.ace-fb-comments');
	ace.fb.ready(function(){
		var href = 'http://'+window.location.host+window.location.pathname;
		$fbComments.html('<div class="fb-comments fb-comments-responsive" data-href="'+href+'" data-width="100%" data-numposts="5" data-colorscheme="light"></div>')
			.css('display','block')
		FB.XFBML.parse($fbComments[0]);
	});
});

</script>

<style type="text/css">
body {
	font-family: Helvetica,arial,sans-serif;
}
a {
	text-decoration: none;
	color: #9e6618;
}
a:hover {
	text-decoration: underline;
}
.page {
	width: 960px;
	margin: 20px auto;
}
.page-title {
	padding: 1em 0 1.4em 0;
	font-size: 25px;
	text-transform: uppercase;
}
.page-title span:nth-child(1) {
	font-weight: 300;
	letter-spacing: 0.19em;
}
.page-title span:nth-child(2) {
	font-weight: 500;
	letter-spacing: 0.15em;
}
h3 {
	font-size: 15px;
	text-transform: uppercase;
	letter-spacing: 0.1em;
	margin: 2em 0 1em 0;
	border-bottom: 1px solid #ccc;
	padding-bottom: 0.3em;
}
h3 span {
	font-size: 12px;
	text-transform: none;
}
.instagram-container {
	float: left;
	width: 660px;
}
.twitter-container {
	float: left;
	width: 300px;
}
.carousel-container {
	padding: 16px 0 0 0;
	margin-bottom: 14px;
}
.ace-twitter.type-featured .ace-twitter-wrap {
	background-position: 933px 39px;
}
.ace-example {
	margin-bottom: 14px;
}
.ace-double .ace-example {
	float: left;
	width: 48%;
}
.ace-double .ace-example:nth-child(2) {
	float: right;
}
div.ace-smile {
	padding: 10px;
	display: inline-block; /* for tooltip */
}
.ace-smile-face {
	cursor: pointer;
	display: inline-block;
	white-space: nowrap;
	overflow: hidden;
}

.ace-fb-comments {
	display: none;
}
</style>

</head>
<body>

<div class="page">

	<h1 class="page-title"><span>Para</span><span>digm</span></h1>

	<div class="instagram-container">
		<!--<script type="text/ace-instagram">{
			query: 'users/227962011/media/recent'
		}</script>-->
		<script type="text/ace-instagram">{
			query: 'users/%userId%/media/recent'
			,fetch: {
				userId: ['users/search?q=jewelmint',function(res){
					return res.data[0].id;
				}]
			}
		}</script>
	</div>

	<div class="twitter-container">
		<script type="text/ace-twitter">{
			screenName: 'luckymagazine'
			,numGet: 10
			,numShow: 4
			,scroll: 'y'
			,type: 'sidebar'
		}</script>
	</div>

	<div class="clear">&nbsp;</div>

	<div class="carousel-container">
		<script type="text/ace-carousel">{
				imgs: [
					'http://thecenteroncentral.com/wp-content/uploads/2014/01/Fashion-Design-camp.jpg'
					,'http://b-i.forbesimg.com/geristengel/files/2013/05/i-ella-fashion-closet.jpg'
					,'http://thechrisellefactor.com/wp-content/uploads/2012/04/Chanel_Wallet_Chriselle1.jpg'
					,'http://thechrisellefactor.com/wp-content/uploads/2012/09/Peplum_Leather_zara_top_Valentino_rockstud_leopard_shoes_Purse_chriselle_Lim_5.jpg'
					,'http://glambistro.com/wp-content/uploads/2013/10/summer-arm-candy.jpg'
					,'http://www.mychicadventure.com/wp-content/uploads/2014/05/boho-chic-2.jpg'
				]
				,dims: '195x148'
				,speed: 300
				,shadbox: true
		}</script>
	</div>

	<div class="twitter-horizontal-container">
		<script type="text/ace-twitter">{
			screenName: 'luckymagazine'
			,numGet: 10
			,numShow: 1
			,scroll: 'x'
			,type: 'featured'
		}</script>
	</div>

	<div class="ace-double">
		<div class="ace-example">
			<h3>ace.tooltip():</h3>
			<a id="ace-tooltip-example" href="#">Hover me!</a>
		</div>
		<div class="ace-example">
			<h3>ace.pop():</h3>
			<a id="ace-pop-example" href="#">Click on me!</a>
		</div>
		<div class="clear">&nbsp;</div>
	</div>

	<div class="ace-example">
		<h3>Ace API Example <span>(/ace/api/demo/smile)</span>:</h3>
		<div class="ace-smile" x-times="11" style="text-align:center;display:block;"></div>
	</div>

	<div class="ace-example">
		<h3>Ace PHP Utility Methods <span>(displayed using Ace::varDump)</span>:</h3>
		<?php
			$c = new ReflectionClass('\ace\Ace');
			$methods = array();
			$ignore = array('loadConfig',);
			$pretty = array();
			//Ace::varDump($c->getDocComment());
			foreach ($c->getMethods() as $v) {
				if ($v->name[0] == '_' || !$v->isPublic() || in_array($v->name,$ignore))
					continue;
				$m = array();
				$comment = trim(preg_replace('/(\/\*)|(\*\/)|\*/','',$v->getDocComment()));
				if ($comment)
					$m['comment'] = $comment;
				$methods[$v->name] = $m;
				$pretty[] = 'Ace::'.($v->name . ($comment ? "\n\t$comment" : ''));
			}
			Ace::varDump($methods);
			//if (!empty($_GET['debug'])) echo "\n\n".implode("\n",$pretty)."\n\n";
		?>
	</div>

	<div class="ace-fb-comments"></div>

</div>

<?php include WEBROOT.'/global-footer.php'; ?>

</body>
</html>